### DISCUSSION TITLE
Stop reindexing translations on every search input change

### DISCUSSION THREAD

**Discussion by SharakPL** [PrestaShop/PrestaShop#35557] [Ideas]
I open Translations to search specific strings and translate them, but the page loads all translations at the start and on every search string change. This is completely useless and makes whole process much slower than it should be. It affects even the production server, though is slow enough to notice on windows dev server. It's been like that from the start of this new translations system of PS1.7 and it's still the same on PS8. Video says it all:

https://user-images.githubusercontent.com/5007456/231437379-ed28336f-2bb1-477c-896f-e39805bf96e0.mp4

### Expected behavior

It should **NEVER** load all translations, because nobody ever needs it. Most of the time you visit Translations to fix just a few of missing ones or change a few of existing ones. Not to change every translation there is. There should be separate button for rebuilding index, just like it is for clearing cache.

### PrestaShop version(s) where the bug happened

1.7, 8.x, develop

### PHP version(s) where the bug happened

7.4, 8.0 8.1

**Comment by SharakPL** [comment]
@AureRita it's not. The issue is about fetching all translations many times on the same page, which is completely unnecessary and hurts performance. Slow wampserver configuration just helps point out where is the problem.

**Comment by AaronTajriba** [comment]
@AureRita We're using an good live server and an upgrade the entire translations page takes longer than 2 min to load! We're using 10 different languages and have thousands translations PER language! This is an big issue. it takes 2 min to load initially and then when you navigate it also takes 2 more minutes (if it loads at all).

@SharakPL Did you found any fix for this issue by any chance?

**Reply by SharakPL** [reply]
No, I didn't. I don't know much about this translations system and what can be done to improve it, but improvements are very much needed. After all these years and with all its flaws I still think the system used up to PS 1.6 was better than what we have now.

**Comment by adviva** [comment]
We are facing kinda the same problem affecting our translation back office page. We have a **HUGE** “Waiting for server response” time (even 27 seconds load!) which we can’t explain. We have done many Prestashop sites and this is the first time facing such a problem. (v8.1.4). We have many sites inside the same server performing very well on that page.

For example, in here “_/api/translations/it-IT/AdminNavigationNotification_”, does have 1000 expressions on it. which 404 are missing. I don’t think we are even talking about big numbers here then.

![scsccscs](https://github.com/PrestaShop/PrestaShop/assets/26090185/22ae048d-add3-4c03-ad03-ca5080cc76b6)
![Screenshot 2024-03-04 at 18 19 17](https://github.com/PrestaShop/PrestaShop/assets/26090185/7576255c-2120-4708-b951-b8ef3b1a88dc)

I think it is anyway related to https://github.com/PrestaShop/PrestaShop/issues/34604

**Comment by Slews** [comment]
I'm quite surprised that this subject seems to be neglected because it's indeed a significant issue.

The thing is, the more modules you have on your shop, the more the translations interface struggles. I think this explains why the PrestaShop Core team did not see that and why nobody seems to care about it. The testing context is too limited and lacks a vision for shops with a large number of data/modules.

To help others understand the depth of the issue, I'll explain our case with some metrics. I think it will help to see that this is really an issue.

For those looking for a patch to improve the situation a bit (just a bit, unfortunately), I will share our current "solution" in a follow-up comment.

### Our Context:

- We are currently upgrading several shops for many clients from PS 1.7.6.5 to PS 8.1.7.
- Each shop has around 150 modules.
- Each shop operates in two or three languages.

Here's what I've observed by comparing the previous version of the shops (PS 1.7.6.5) and the latest version we're working on (PS 8.1.7).

### Observations:
																	  
For both versions, the way the page works hasn't fundamentally changed. First, the page loads almost empty, then a first Ajax request is sent to load the domain tree, followed by a second Ajax request to load the main form (with the translations for the current/default domain).

![image](https://github.com/user-attachments/assets/99392533-3da9-4c3d-804c-ac3bcc241479)

### Measurements:

Note that all measurements were made on the same server to avoid any network or server configuration bias. Also, the data are averages, the important part is not the value itself, but the proportion between cases.

For the 1.7.6.5 version, the Ajax requests execute in:

- Domain tree request:  3.150s
- Content request: 0.417s

For the 8.1.7 version, the Ajax requests execute in:

- Domain tree request: 15.990s
- Content request: 15.628s (almost 40 times slower !)

The difference is very significant.

This shows that just by upgrading PrestaShop, what was largely acceptable before becomes mostly unusable.

**Reply by Slews** [reply]
### A way to improve a bit:

Please note that the following is not a permanent solution but a temporary patch to improve our situation. It is not clean and lacks tests to ensure no side effects.

### Overview of Changes

1. Focus on TranslationController::listDomainTranslationAction:
The main difference between versions 1.7.6.5 and 8.1.7 lies in the $translationService->listDomainTranslation method, which now uses ProviderDefinition as the first parameter.
	
2. Objective:
To mitigate the impact of having many modules, we targeted methods dealing with module translations within the **ThemeCatalogueLayersProvider** class.
	
### Modifications Made in `ThemeCatalogueLayersProvider`
	
1. `getModulesFromTranslations` Method:

**Original method:**
```
private function getModulesFromTranslations(string $locale): array
{
	$modules = [];

	$catalogue = $this->getDefaultCatalogue($locale);
	foreach ($catalogue->getDomains() as $domain) {
		if (preg_match('/^Modules([A-Z]([^A-Z]+))/', $domain, $matches)) {
			$modules[] = strtolower($matches[1]);
		}
	}

	return $modules;
}
```

**Problem:**
As said earlier, the problem increases with the number of modules. And here, a single module name could be added several time since no test is there to avoid duplicates.

**Modified method:**
```
private function getModulesFromTranslations(string $locale): array
{
    $modules = [];
    $catalogue = $this->getDefaultCatalogue($locale);
    foreach ($catalogue->getDomains() as $domain) {
        if (preg_match('/^Modules([A-Z]([^A-Z]+))/', $domain, $matches)) {
            $moduleName = strtolower($matches[1]);
            if (!in_array($moduleName, $modules)) {
                $modules[] = $moduleName;
            }
        }
    }
    return $modules;
}
```

**Result:**
Requests are significantly faster (time reduced by over half):
Domain tree request: 6.230s
Content for domain Admin.Global: 6.050s


2. `getFileTranslatedCatalogue` Method:

**Original method:**
```
public function getFileTranslatedCatalogue(string $locale): MessageCatalogue
{
	// load front office catalogue
	$coreCatalogue = $this->getCoreCatalogue($locale);
	$coreCatalogue->addCatalogue($this->getModulesTranslations($locale));

	try {
		$fileTranslatedCatalogue = (new FileTranslatedCatalogueFinder(
			$this->getResourceDirectory(),
			['*']
		))
			->getCatalogue($locale);

		// overwrite with the theme's own catalogue
		$coreCatalogue->addCatalogue($fileTranslatedCatalogue);
	} catch (TranslationFilesNotFoundException $e) {
		// No translation file was found in the theme, we keep using those from the core
	}

	return $coreCatalogue;
}
```

**Problem:**
The $coreCatalogue->addCatalogue($this->getModulesTranslations($locale)); line is always executed, even if we are displaying non module-related translation.
We need a test to avoid that.

**Modified method:**
```
public function getFileTranslatedCatalogue(string $locale): MessageCatalogue
{
    $coreCatalogue = $this->getCoreCatalogue($locale);
    $domain = \PrestaShop\PrestaShop\Adapter\SymfonyContainer::getInstance()
                    ->get('request_stack')
                    ->getCurrentRequest()
                    ->get('domain');

    if (empty($domain) || strpos($domain, 'Modules') !== false) {
        $coreCatalogue->addCatalogue($this->getModulesTranslations($locale));
    }

    try {
        $fileTranslatedCatalogue = (new FileTranslatedCatalogueFinder(
            $this->getResourceDirectory(),
            ['*']
        ))->getCatalogue($locale);

        $coreCatalogue->addCatalogue($fileTranslatedCatalogue);
    } catch (TranslationFilesNotFoundException $e) {
        // No translation file was found in the theme, we keep using those from the core
    }

    return $coreCatalogue;
}
```

The test is not great (just a POC), the idea was to differenciate the first request and the second, and then to check if the called domain is module-related.

**Result:**
Non-module-related translation domains load faster (2.150s).

### Conclusion
With these changes in the ThemeCatalogueLayersProvider class, page load times have improved from 32 seconds to 8-12 seconds. While not ideal, this brings us from unusable to painfully usable.

**Comment by elyesnaouar** [comment]
I checked the proposition for PS9.x. I confirm there is a performance issue by loading unnecessary pages as can be seen in the screenshot below 
<img width="1435" height="505" alt="image" src="https://github.com/user-attachments/assets/f6885680-73f2-40a6-8f39-9578628b10e4" />

<img width="1435" height="296" alt="image" src="https://github.com/user-attachments/assets/e2f3c2c0-2afb-438b-b57b-3f611ddd2830" />
<img width="430" height="148" alt="image" src="https://github.com/user-attachments/assets/2702374a-e9a0-4546-83a0-6e44b759d0f4" />

may be load lazing or pagination should be used to reduce loading time

**Reply by SharakPL** [reply]
Yeah, even on an optimized Linux server it's slow. Much slower than it should be.
